include/albatross/models/sparse_gp.h

```diff
@@ -124,43 +124,61 @@ class SparseGaussianProcessRegression
-  auto _fit_impl(const std::vector<FeatureType> &features,
-                 const MarginalDistribution &targets) const {
-    static_assert(std::is_same<IndexingFunction, LeaveOneOut>::value,
-                  "Only FITC is currently implemented, but by allowing "
-                  "arbitrary groups this could turn into PITC");
+  auto _fit_impl(const std::vector<FeatureType> &out_of_order_features,
+                 const MarginalDistribution &out_of_order_targets) const {
+
+    const auto indexer =
+        independent_group_indexing_function(out_of_order_features);
+
+    std::vector<std::size_t> reordered_inds;
+    BlockDiagonal K_ff;
+    for (const auto &pair : indexer) {
+      reordered_inds.insert(reordered_inds.end(), pair.second.begin(),
+                            pair.second.end());
+      auto subset_features = subset(out_of_order_features, pair.second);
+      K_ff.blocks.emplace_back(this->covariance_function_(subset_features));
+      if (out_of_order_targets.has_covariance()) {
+        K_ff.blocks.back().diagonal() +=
+            subset(out_of_order_targets.covariance.diagonal(), pair.second);
+      }
+    }
+
+    const auto features = subset(out_of_order_features, reordered_inds);
+    const auto targets = subset(out_of_order_targets, reordered_inds);
 
     // Determine the set of inducing points, u.
-    auto u = inducing_point_strategy(features);
+    const auto u = inducing_point_strategy(features);
 
-    Eigen::Index n = targets.mean.size();
     Eigen::Index m = static_cast<Eigen::Index>(u.size());
 
-    // Create the covariance matrices we need.
-    Eigen::VectorXd K_ff_diag = this->covariance_function_.diagonal(features);
-    if (targets.has_covariance()) {
-      K_ff_diag += targets.covariance.diagonal();
-    }
-    Eigen::MatrixXd K_fu = this->covariance_function_(features, u);
-    Eigen::MatrixXd K_uu = this->covariance_function_(u);
+    const Eigen::MatrixXd K_fu = this->covariance_function_(features, u);
+    const Eigen::MatrixXd K_uu = this->covariance_function_(u);
 
-    auto K_uu_llt = K_uu.llt();
+    const auto K_uu_llt = K_uu.llt();
     // P is such that:
     //     Q_ff = K_fu K_uu^-1 K_uf
     //          = K_fu L^-T L^-1 K_uf
     //          = P^T P
-    Eigen::MatrixXd P = K_uu_llt.matrixL().solve(K_fu.transpose());
+    const Eigen::MatrixXd P = K_uu_llt.matrixL().solve(K_fu.transpose());
 
-    // Efficiently compute the diagonal diag[Q_ff].
-    Eigen::VectorXd Q_ff_diag = P.colwise().squaredNorm();
+    BlockDiagonal Q_ff;
+    Eigen::Index i = 0;
+    for (const auto &pair : indexer) {
+      Eigen::Index cols = static_cast<Eigen::Index>(pair.second.size());
+      auto P_cols = P.block(0, i, P.rows(), cols);
+      Q_ff.blocks.emplace_back(P_cols.transpose() * P_cols);
+      i += cols;
+    }
 
-    Eigen::VectorXd A = K_ff_diag - Q_ff_diag;
+    auto A = K_ff - Q_ff;
 
-    if (A.minCoeff() < 1e-6) {
+    if (A.diagonal().minCoeff() < 1e-6) {
```

Comment:
Magic numbers. Does this lower bound make sense for all GPs, or just the OE use case?